### PR TITLE
CloudFoundry Executor with runner manager using cf push

### PR DESCRIPTION
**Pull request by pauldoomgov**
Hopping off work by @mogul I am trying a slightly different executor approach that would allow using images chosen on-the-fly and managing the lifecycle of the child runner directly.

Startup times for a do-nothing Cloud.gov app seem to run under 30 seconds with teardown under 10 seconds. Using a nearby OCI image might bring that down more.

This is PR gets basic functionality in place. A complete run with the following `.gitlab-ci.yml` takes about 2 minutes and 40 seconds. Follow on PRs can work on shrinking this down.

~~~yaml
variables:
  # DELETE JOB LOGS AFTER TURNING THIS OFF
  CI_DEBUG_TRACE: "true"
stages:
- build
- test
- deploy
build-job:
  stage: build
  script:
  - echo "Compiling the code..."
  - echo "Compile complete."
unit-test-job:
  stage: test
  script:
  - echo "Running unit tests... This will take about 10 seconds."
  - sleep 10
  - echo "Code coverage is 0%, loser"
lint-test-job:
  stage: test
  script:
  - echo "Linting code... This will take about 10 seconds."
  - sleep 10
  - echo "Your code is most foul."
deploy-job:
  environment: production
  stage: deploy
  script:
  - echo "Deploying application..."
  - echo "Application deployed - Service is down."
 ~~~

image

### CODE REVIEW

**Comment by pauldoomgov** [comment]
Note https://github.com/GSA-TTS/devtools-program/issues/38#issuecomment-2043681610 - The need for "helper" components will make this more complicated. Will try the approach of firing up a `gitlab-runner-helper` sidecar.

**Comment by mogul** [comment]
"Sidecar" in the Cloud Foundry world means something different than in the k8s world, and I'm about 90% certain that will mean the sidecar route is not viable, though that remaining 10% is reason to still poke a toe down that path.

In CF, a Docker image is internally converted into a droplet (a CF-flavored image format, and what results after you combine app code + buildpacks + stack during the buildpack lifecycle), and then runs in a container just like any other CF app, where there's a main process which is started by the platform and then monitored for exit. A "sidecar" is just another command that the platform starts and monitors _inside the same container as the main process_. In other words, the sidecar is not a separate container at all, and you need to make sure that everything it needs is in the droplet before you go adding an entry in the app's manifest for it. (For example, a sidecar process might be a log-shipper, or a mmonit process, or the "real" app when the main app is an nginx proxy, etc.) It doesn't help you with orchestrating two containers from two separate OCI images.

That said, you may be able to have your worker image derive `FROM` the helper image, in which case, yeah, then it'll work. But I got the sense from what you wrote in https://github.com/GSA-TTS/devtools-program/issues/38#issuecomment-2043681610 that LXC went that route and you wanted to avoid it...?

**Comment by pauldoomgov** [comment]
Thanks for the great clarification, @mogul! My goal was to keep the container filesystem clean, so same-chroot/multi-process sidecar fails on that count.

Pivoting to the approach the LXC example used.

**Comment by mogul** [comment]
> you may be able to have your worker image derive FROM the helper image

I thought about it and that's not a good plan... You should add whatever's in the helper image _on top_ of the upstream distro FROMs that your workers are using. For compliance reasons, you likely already need to implement local scanning of the upstream images before using them, so you may as well just add the helper content into your own images. See [this example](https://github.com/GSA-TTS/FAC/actions/workflows/pull-containers-and-push-to-ghcr.yml) and [this one](https://github.com/GSA-TTS/FAC/blob/main/.github/workflows/trivy.yml).

**Comment by pauldoomgov** [comment]
`prepare.sh` is working.

Now we are dying on this error: https://docs.gitlab.com/runner/shells/index.html#shell-profile-loading  That is a tomorrow Paul problem.

File: README.md [hunk 1]
```diff
@@ -20,21 +20,39 @@ Currently this just eases the task of deploying the `gitlab-runner` binary and r
...
     cf push --vars-file vars.yml
     ```
-6. Check to see that the runner has registered itself in GitLab
+7. Check to see that the runner has registered itself in GitLab
 
 ## TODO
 
 - Add a `gitlab-worker` app
-- Write a [custom `gitlab-runner` Executor](https://docs.gitlab.com/runner/executors/custom.html) that uses `cf run-task gitlab-runner -w -c $@` (refer to [the libvirt example](https://docs.gitlab.com/runner/executors/custom_examples/libvirt.html))
+- Write a [custom `gitlab-runner` Executor](https://docs.gitlab.com/runner/executors/custom.html) that uses `cf ssh gitlab-worker $@` (refer to [the libvirt example](https://docs.gitlab.com/runner/executors/custom_examples/libvirt.html))
```
Comment (mogul) [inline README.md:46]: This is done now, right? We should remove it.

**Review by mogul — CHANGES_REQUESTED** [review]
Other than removing the one leftover TODO, I'd also like to see some instructions for verifying that it's working from the Gitlab side, please... Is there a simple repository CI/CF configuration that this could be tested with, post registration? Where would I look to verify that it's doing what's expected?

**Comment by mogul** [comment]
Oh, I see that you added a verification config in the top post! I think that you just need to move that into the tree, so it's not only documented here on the PR.

File: README.md (hunk 1 above)
Comment (pauldoomgov) [reply README.md]: Yes! And I refined the TODO to include what I had scattered in code.

**Review by mogul — APPROVED** [review]

**Pull request merged**